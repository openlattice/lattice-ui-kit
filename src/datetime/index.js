export { default as DatePicker } from './src/components/DatePicker';
export { default as TimePicker } from './src/components/TimePicker';

export { default as MaterialDatePicker } from './src/components/MaterialDatePicker';
export { default as MaterialTimePicker } from './src/components/MaterialTimePicker';
export { default as DateTimePicker } from './src/components/DateTimePicker';
