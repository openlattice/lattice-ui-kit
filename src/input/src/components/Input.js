import styled from 'styled-components';
import { NEUTRALS, PURPLES, RED_1 } from '../../../colors';

const Input = styled.input`
  display: flex;
  flex: 0 1 auto;
  box-sizing: border-box;
  background: ${NEUTRALS[8]} url(${props => (props.icon ? props.icon : null)}) no-repeat center right 10px;
  border-radius: 3px;
  border: solid 1px ${props => (props.invalid ? RED_1 : NEUTRALS[4])};
  box-shadow: 0;
  color: ${NEUTRALS[0]};
  font-size: 14px;
  height: 39px;
  line-height: 19px;
  padding: ${props => (props.icon ? '0px 32px 0px 10px' : '0px 10px')};
  text-overflow: ellipsis;
  width: 100%;
  transition: background-color 0.2s ease-in-out, border-color 0.2s ease-in-out;

  :hover {
    background-color: ${NEUTRALS[6]};
  }

  :focus {
    border: solid 1px ${PURPLES[1]};
    background-color: white;
    outline: none;
  }

  :disabled {
    background-color: ${NEUTRALS[8]};
    color: ${NEUTRALS[1]};
    cursor: not-allowed;
  }
`;

export default Input;
