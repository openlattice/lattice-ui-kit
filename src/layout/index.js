import AppContainerWrapper from './src/components/App/AppContainerWrapper';
import AppContentWrapper from './src/components/App/AppContentWrapper';
import AppHeaderWrapper from './src/components/App/AppHeaderWrapper';
import AppNavigationWrapper from './src/components/App/AppNavigationWrapper';
import Card from './src/components/Card/Card';
import CardHeader from './src/components/CardHeader/CardHeader';
import CardSegment from './src/components/CardSegment/CardSegment';
import CardStack from './src/components/CardStack/CardStack';

export {
  AppContainerWrapper,
  AppContentWrapper,
  AppHeaderWrapper,
  AppNavigationWrapper,
  Card,
  CardHeader,
  CardSegment,
  CardStack,
};
