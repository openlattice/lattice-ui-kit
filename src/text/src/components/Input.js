// @flow

import styled from 'styled-components';
import inputStyles from './styled/inputStyles';

const Input = styled.input`
  ${inputStyles}
`;

export default Input;
