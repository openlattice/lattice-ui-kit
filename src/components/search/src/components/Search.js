// @flow
import React, { Component } from 'react';
import isFunction from 'lodash/isFunction';
import { Map, List } from 'immutable';
import type { ComponentType, Node } from 'react';

import DefaultResultComponent from './Result';
import DefaultSearchResults from './SearchResults';
import DatePicker from '../../../../datetime/src/components/DatePicker';
import Label from '../../../../label';
import { Button } from '../../../../button';
import { Input } from '../../../../text';
import { InputGrid, Title } from './styled/StyledSearchComponents';
import { CheckboxSelect } from '../../../../select';
import { Card, CardSegment, CardStack } from '../../../../layout';

import type { SearchResultsProps } from './SearchResults';
import type { ResultProps } from './Result';
import type { SearchFieldDefinition, FilterFieldDefinition } from '../../types';
import type { ReactSelectEvent, ReactSelectOption, ReactSelectValue } from '../../../../select/types';

type Props = {
  className ? :string;
  hasSearched ? :boolean;
  isLoading :any;
  filterFields ? :FilterFieldDefinition[];
  onResultClick ? :(result :Map) => void;
  onSearch :(searchFieldValues :Map) => void;
  resultComponent ? :ComponentType<ResultProps>;
  resultLabels ? :Map;
  searchFields ? :SearchFieldDefinition[];
  searchResults ? :List<Map>;
  searchResultsComponent ? :ComponentType<SearchResultsProps>;
  title :string;
};

type State = {
  searchFieldValues :Map;
  filterFieldValues :Map;
};

class Search extends Component<Props, State> {

  static defaultProps = {
    className: undefined,
    filterFields: [],
    hasSearched: false,
    onResultClick: undefined,
    resultComponent: DefaultResultComponent,
    resultLabels: undefined,
    searchFields: [
      {
        id: 'firstname',
        label: 'First Name'
      },
      {
        id: 'lastname',
        label: 'Last Name'
      },
      {
        id: 'dob',
        label: 'Date of Birth',
        type: 'date'
      },
    ],
    searchResults: List(),
    searchResultsComponent: DefaultSearchResults,
  }

  constructor(props :Props) {
    super(props);

    const searchFieldValues = Map().withMutations((map :Map) => {
      // $FlowFixMe optional not recognizing defaultProps
      props.searchFields.forEach((field :SearchFieldDefinition) => map.set(field.id, ''));
    });

    this.state = {
      searchFieldValues,
      filterFieldValues: Map()
    };
  }

  handleOnChangeInput = (e :SyntheticEvent<HTMLInputElement>) => {
    const { searchFieldValues } = this.state;
    const { value = '', name } = e.currentTarget;
    this.setState({
      searchFieldValues: searchFieldValues.set(name, value),
    });
  }

  getOnChangeDate = (name :string) => (date :string) => {
    const { searchFieldValues } = this.state;
    this.setState({
      searchFieldValues: searchFieldValues.set(name, date)
    });
  }

  handleOnClickSearchButton = (e :SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    const { onSearch } = this.props;
    const { searchFieldValues } = this.state;
    if (isFunction(onSearch)) {
      onSearch(searchFieldValues);
    }
  }

  handleOnChangeFilter = (value ? :ReactSelectValue, event :ReactSelectEvent) => {
    const { filterFieldValues } = this.state;
    const { name } = event;
    this.setState({
      filterFieldValues: filterFieldValues.set(name, value)
    });
  }

  renderFilteredSearchResults = () :Node => {
    const {
      hasSearched,
      isLoading,
      filterFields,
      onResultClick,
      resultComponent,
      resultLabels,
      searchResults,
      searchResultsComponent: SearchResultsComponent,
    } = this.props;
    const { filterFieldValues } = this.state;

    let filteredResults = List();
    if (searchResults && List.isList(searchResults)) {

      filteredResults = searchResults.filter((searchResult :Map) => (
        // accumulate matchAllFilters from all filterFieldValue keys where
        filterFieldValues.reduce((matchAllFilters :boolean, selectedValue :ReactSelectValue, fieldName :string) => {

          let filterFieldDef :?FilterFieldDefinition;
          if (filterFields && filterFields.length) {
            filterFieldDef = filterFields.find(
              (definition :FilterFieldDefinition) => (definition.id === fieldName)
            );
          }

          // there is matching filterFieldDef
          if (filterFieldDef) {
            // whose filterCallback yields truthy when provided the searchResult and selectedValue(s)
            return matchAllFilters && filterFieldDef.filterCallback(searchResult, selectedValue);
          }

          return matchAllFilters;
        }, true)
      ));

    }

    if (SearchResultsComponent) {
      return (
        <SearchResultsComponent
            hasSearched={hasSearched}
            isLoading={isLoading}
            onResultClick={onResultClick}
            resultComponent={resultComponent}
            resultLabels={resultLabels}
            results={filteredResults} />
      );
    }

    return null;
  }

  renderSearchFieldsSegment = () => {

    const { searchFields, title, isLoading } = this.props;
    const { searchFieldValues } = this.state;

    // $FlowFixMe optional not recognizing defaultProps
    const searchFieldComponents = searchFields.map((field :SearchFieldDefinition) => {

      let fieldComponent = (
        <Input
            id={`luk-search-${field.id}`}
            name={field.id}
            onChange={this.handleOnChangeInput}
            value={searchFieldValues.get(field.id)} />
      );

      // TODO: More flexible method for handling checkboxes, dropdowns, etc.
      if (field.type === 'date') {
        fieldComponent = (
          <DatePicker
              id={`luk-search-${field.id}`}
              name={field.id}
              onChange={this.getOnChangeDate(field.id)}
              value={searchFieldValues.get(field.id)} />
        );
      }

      return (
        <div key={`luk-search-key-${field.id}`}>
          <Label htmlFor={`luk-search-${field.id}`}>
            {field.label}
          </Label>
          {fieldComponent}
        </div>
      );
    });

    return (
      <CardSegment vertical>
        { title && <Title>{title}</Title> }
        <form>
          {/* $FlowFixMe optional not recognizing defaultProps */}
          <InputGrid columns={searchFields.length + 1} align="flex-start">
            {searchFieldComponents}
            <div>
              <Label stealth>Submit</Label>
              <Button
                  type="submit"
                  fullWidth
                  isLoading={isLoading}
                  mode="primary"
                  onClick={this.handleOnClickSearchButton}>
                Search
              </Button>
            </div>
          </InputGrid>
        </form>
      </CardSegment>
    );
  }

  renderFiltersSegment = () => {

    const { filterFields } = this.props;
    if (Array.isArray(filterFields) && filterFields.length) {
      const filterFieldComponents = filterFields.map((filter :FilterFieldDefinition) => {
        const options :ReactSelectOption[] = filter.options.map((v) => ({ label: v, value: v }));
        return (
          <div key={`luk-filter-key-${filter.id}`}>
            <Label
                bold
                key={filter.id}
                htmlFor={`luk-filter-${filter.id}`}>
              {filter.label}
            </Label>
            <CheckboxSelect
                name={filter.id}
                id={`luk-filter-${filter.id}`}
                borderless
                placeholder="Add filter"
                onChange={this.handleOnChangeFilter}
                options={options} />
          </div>
        );
      });

      return (
        <CardSegment vertical padding="sm">
          <InputGrid columns={filterFields.length} align="flex-start">
            {filterFieldComponents}
          </InputGrid>
        </CardSegment>
      );
    }

    return null;
  }

  render() {
    const { className } = this.props;
    return (
      <CardStack className={className}>
        <Card>
          { this.renderSearchFieldsSegment() }
          { this.renderFiltersSegment() }
        </Card>
        { this.renderFilteredSearchResults() }
      </CardStack>
    );
  }
}

export default Search;
export type { Props as SearchProps };
