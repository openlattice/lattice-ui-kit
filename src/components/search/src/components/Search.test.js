import React from 'react';
import toJson from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';

import Search from './Search';
import Input from '../../../../input';
import { Card } from '../../../../layout';
import { DatePicker } from '../../../../datetime';

describe('Search', () => {

  describe('snapshots', () => {

    test('should match snapshot', () => {
      const wrapper = mount(<Search />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

  });

  describe('render', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(<Search />);
    });

    test('render Card', () => {
      expect(wrapper.find(Card)).toHaveLength(1);
    });

    test('default search fields', () => {
      expect(wrapper.find(Input)).toHaveLength(2);
      expect(wrapper.find(DatePicker)).toHaveLength(1);
    });

    test('search button', () => {
      expect(wrapper.find('button')).toHaveLength(1);
    });

  });

  describe('onSearch()', () => {
    test('should invoke onSearch', () => {
      const mockOnSearch = jest.fn();
      const wrapper = mount(<Search onSearch={mockOnSearch} />);
      wrapper.find('button').simulate('click');
      expect(mockOnSearch).toHaveBeenCalledTimes(1);
    });
  });

  describe('searchFieldValues', () => {
    test('change input', () => {
      const wrapper = shallow(<Search />);
      const firstNameInput = wrapper.find('Input[name="firstname"]');

      expect(firstNameInput).toHaveLength(1);
      expect(firstNameInput.type()).toEqual(Input);

      const changeValue = 'OpenLattice';
      const initialSearchFieldValues = wrapper.state('searchFieldValues');
      const expectedSearchFieldValues = initialSearchFieldValues.set('firstname', changeValue);

      firstNameInput.simulate('change', { currentTarget: { value: changeValue, name: 'firstname' } });

      const actualSearchFieldValues = wrapper.state('searchFieldValues');
      expect(actualSearchFieldValues).toStrictEqual(expectedSearchFieldValues);

    });

    test('change date', () => {
      const wrapper = shallow(<Search />);
      const datePicker = wrapper.find(DatePicker);

      expect(datePicker).toHaveLength(1);
      expect(datePicker.type()).toEqual(DatePicker);

      const changeValue = '1970-01-01';
      const initialSearchFieldValues = wrapper.state('searchFieldValues');
      const expectedSearchFieldValues = initialSearchFieldValues.set('dob', changeValue);

      datePicker.simulate('change', changeValue);

      const actualSearchFieldValues = wrapper.state('searchFieldValues');
      expect(actualSearchFieldValues).toStrictEqual(expectedSearchFieldValues);

    });
  });

});
