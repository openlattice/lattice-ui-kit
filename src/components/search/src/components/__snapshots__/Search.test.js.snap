// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search snapshots should match snapshot 1`] = `
<Card>
  <CardSegment
    vertical={true}
  >
    <form>
      <InputGrid>
        <div
          key="luk-search-key-firstname"
        >
          <Label
            htmlFor="luk-search-firstname"
          >
            First Name
          </Label>
          <Input
            name="firstname"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          key="luk-search-key-lastname"
        >
          <Label
            htmlFor="luk-search-lastname"
          >
            Last Name
          </Label>
          <Input
            name="lastname"
            onChange={[Function]}
            value=""
          />
        </div>
        <div
          key="luk-search-key-dob"
        >
          <Label
            htmlFor="luk-search-dob"
          >
            Date of Birth
          </Label>
          <LatticeWrapper
            name="dob"
            onChange={[Function]}
            value=""
          />
        </div>
        <Button
          className=""
          disabled={false}
          mode="primary"
          onClick={[Function]}
          type="submit"
        >
          Search
        </Button>
      </InputGrid>
    </form>
  </CardSegment>
  <CardSegment
    padding="md"
  >
    Filters
  </CardSegment>
</Card>
`;
